blockchain, netsync: implement a complete headers-first download during ibd

kcalvinalvin opened the pull request:
## Change Description
Right now the headers-first download is only based off of the checkpoints and is thus limited to the last checkpoint.
The newly implemented headers-first download will always download headers-first and will validate them to see if the headers connect and have proper proof of work.

Then the block download will be based off of the verified headers. This now eliminates any potential downloading of txs or orphan blocks during ibd. It also makes future parallel block download much better as a parallel block download can only happen for blocks we already have headers for.

~~It's not yet put into the code yet but this allows the node to also receive block headers instead of invs during block propagation.~~ I'll do this in a follow up later.

## Steps to Test

- Perform ibd on any of the networks and see that the node syncs up fine.
- Check that the newly added function tests are correct.

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change is not [insubstantial](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#substantial-contributions-only). Typo fixes are not accepted to fight bot spam.
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#code-documentation-and-commenting) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#model-git-commit-messages). 
- [ ] Any new logging statements use an appropriate subsystem and logging level.

📝 Please see our [Contribution Guidelines](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

saubyk commented:
cc: @mohamedawnallah for review

Roasbeef commented on blockchain/blockindex.go:
```diff
@@ -46,6 +49,11 @@ func (status blockStatus) HaveData() bool {
 }
 
+// HaveHeader returns whether the header data is stored in the database.
+func (status blockStatus) HaveHeader() bool {
```
Isn't this just the inverse of `HaveData`?

Roasbeef commented on blockchain/process_test.go:
```diff
@@ -0,0 +1,231 @@
+			MerkleRoot: merkle,
+			Bits:       chainParams.PowLimitBits,
+			Timestamp:  ts,
+			Nonce:      0, // To be solved.
```
Superfluous comment.

Roasbeef commented on blockchain/process_test.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	// Create invalid header at the checkpoint.
+	//
```
Style nit: blank line, same below.

Roasbeef commented on blockchain/chain.go:
```diff
@@ -1535,6 +1544,52 @@ func (b *BlockChain) BlockHashByHeight(blockHeight int32) (*chainhash.Hash, erro
+		return false
+	}
+
+	if node.status.KnownInvalid() {
```
```go
return !node.status.KnownInvalid()
```

?

kcalvinalvin replied:
Good catch

Roasbeef commented on blockchain/chain.go:
```diff
@@ -1535,6 +1544,52 @@ func (b *BlockChain) BlockHashByHeight(blockHeight int32) (*chainhash.Hash, erro
+}
+
+// HeaderHashByHeight returns the block header's hash given its height.
+func (b *BlockChain) HeaderHashByHeight(blockHeight int32) (*chainhash.Hash, error) {
```
We should add to the `godoc` here and below that this'll return an error if the height/hash isn't in the best chain.

kcalvinalvin replied:
Will add

Roasbeef commented on netsync/manager.go:
```diff
@@ -1075,99 +1075,40 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
-		// that is already in the database and is only used to ensure
-		// the next header links properly, it must be removed before
-		// fetching the blocks.
-		sm.headerList.Remove(sm.headerList.Front())
```
Can we remove the header list all together now? Or is it still needed for re-orgs?

kcalvinalvin commented on blockchain/process_test.go:
```diff
@@ -0,0 +1,231 @@
+	lastHeader := headers[len(headers)-1]
+	lastHeaderHash := lastHeader.BlockHash()
+	tipNode := chain.bestHeader.Tip()
+	if !tipNode.hash.IsEqual(&lastHeaderHash) {
```
Will change

kcalvinalvin commented on netsync/manager.go:
```diff
@@ -1075,99 +1075,40 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
-				"properly connect to the chain from peer %s "+
-				"-- disconnecting", peer.Addr())
+		_, err := sm.chain.ProcessBlockHeader(
+			blockHeader, blockchain.BFNone, false)
```
Will change

kcalvinalvin commented:
Addressed all the review comments by roasbeef

Roasbeef commented:
Tested in the wild, and does what it says on the tin: 
<details>
```
2025-12-17 18:32:10.885 [INF] SYNC: Downloading headers for blocks 136001 to 2401568 from peer PEER_A:18333
2025-12-17 18:32:10.885 [INF] SYNC: Lost peer PEER_B:18333 (outbound)
2025-12-17 18:32:16.309 [INF] SYNC: New valid peer PEER_C:18333 (outbound) (/Satoshi:27.1.0/)
2025-12-17 18:32:40.885 [INF] SYNC: Downloading headers for blocks 136001 to 4807558 from peer PEER_D:18333
2025-12-17 18:32:40.885 [INF] SYNC: Lost peer PEER_A:18333 (outbound)
2025-12-17 18:32:41.044 [INF] SYNC: New valid peer PEER_E:18333 (outbound) (/Satoshi:28.1.0/)
2025-12-17 18:32:41.135 [INF] SYNC: New valid peer PEER_F:18333 (outbound) (/Satoshi:27.0.0/)
2025-12-17 18:32:46.438 [INF] CHAN: Verified checkpoint at height 200000/block 0000000000287bffd321963ef05feab753ebe274e1d78b2fd4e2bfe9ad3aa6f2
2025-12-17 18:32:51.245 [INF] SYNC: New valid peer PEER_G:18333 (outbound) (/Satoshi:28.1.0/)
2025-12-17 18:32:55.171 [INF] CHAN: Verified checkpoint at height 300001/block 0000000000004829474748f3d1bc8fcf893c88be255e6d7f571c548aff57abf4
2025-12-17 18:33:03.103 [INF] CHAN: Verified checkpoint at height 400002/block 0000000005e2c73b8ecb82ae2dbc2e8274614ebad7172b53528aba7501f5a089
2025-12-17 18:33:10.885 [INF] SYNC: Downloading headers for blocks 490001 to 4807558 from peer PEER_G:18333
2025-12-17 18:33:10.885 [INF] SYNC: Lost peer PEER_D:18333 (outbound)
2025-12-17 18:33:13.135 [INF] CHAN: Verified checkpoint at height 500011/block 00000000000929f63977fbac92ff570a9bd9e7715401ee96f2848f7b07750b02
2025-12-17 18:33:25.990 [INF] CHAN: Verified checkpoint at height 600002/block 000000000001f471389afd6ee94dcace5ccc44adc18e8bff402443f034b07240
2025-12-17 18:33:38.331 [INF] CHAN: Verified checkpoint at height 700000/block 000000000000406178b12a4dea3b27e13b3c4fe4510994fd667d7c1e6a3f4dc1
2025-12-17 18:33:40.884 [INF] SYNC: Downloading headers for blocks 718001 to 4807558 from peer PEER_E:18333
2025-12-17 18:33:40.888 [INF] SYNC: Lost peer PEER_G:18333 (outbound)
2025-12-17 18:33:46.254 [INF] SYNC: New valid peer PEER_H:18333 (outbound) (/Satoshi:25.1.0/)
2025-12-17 18:33:50.353 [INF] CHAN: Verified checkpoint at height 800010/block 000000000017ed35296433190b6829db01e657d80631d43f5983fa403bfdb4c1
2025-12-17 18:34:01.433 [INF] CHAN: Verified checkpoint at height 900000/block 0000000000356f8d8924556e765b7a94aaebc6b5c8685dcfa2b1ee8b41acd89b
2025-12-17 18:34:10.884 [INF] SYNC: Downloading headers for blocks 986001 to 4807558 from peer PEER_E:18333
2025-12-17 18:34:10.884 [INF] SYNC: Lost peer PEER_E:18333 (outbound)
2025-12-17 18:34:10.884 [INF] SYNC: Downloading headers for blocks 986001 to 4807558 from peer PEER_H:18333
2025-12-17 18:34:11.269 [INF] SYNC: New valid peer PEER_I:18333 (outbound) (/Satoshi:25.1.0/)
2025-12-17 18:34:11.270 [INF] SYNC: New valid peer PEER_J:18333 (outbound) (/Satoshi:0.20.1/)
2025-12-17 18:34:13.213 [INF] CHAN: Verified checkpoint at height 1000007/block 00000000001ccb893d8a1f25b70ad173ce955e5f50124261bbbc50379a612ddf
2025-12-17 18:34:16.312 [INF] SYNC: New valid peer PEER_K:18333 (outbound) (/Satoshi:27.0.0/)
2025-12-17 18:34:25.081 [INF] CHAN: Verified checkpoint at height 1100007/block 00000000000abc7b2cd18768ab3dee20857326a818d1946ed6796f42d66dd1e8
```
</details> 



One thing I noticed though is that the peers may D/C us along the way, so we rotate through a few peers to fetch all the headers. This is testnet3 , there's quite a lot of headers (over 4 million). Are we hitting something like default per-peer upload limit for `bitcoind`? 

IIUC this is a stepping stone for multi-peer header download, so perhaps that'll be sorted out as we'll fetch blocks of headers from many peers in parallel.

Abdulkbk commented on netsync/manager_test.go:
```diff
@@ -0,0 +1,239 @@
+				header.BlockHash())
+		}
+	}
+	if err != nil {
```
I think this check is redundant.

Abdulkbk commented on netsync/manager_test.go:
```diff
@@ -297,3 +298,162 @@ func TestFetchHigherPeers(t *testing.T) {
+	return 0
+}
+
+func TestIsInIBDMode(t *testing.T) {
```
godoc

Abdulkbk commented on blockchain/accept.go:
```diff
@@ -92,3 +93,141 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
+	// Orphan headers are not allowed and this function should never be called
+	// with the genesis block.
+	prevHash := &header.PrevBlock
+	prevNode := b.index.LookupNode(prevHash)
```
seems like a duplicate prevnode lookup, we just did it above in the `if node != nil` block. Maybe we should consolidate it?.

kcalvinalvin replied:
Simplified the logic and removed the lookup in the branching in the latest push

yyforyongyu reviewed (commented):
Cool stuff - left a few comments that i think we need to fix before merging.

yyforyongyu commented on blockchain/chain.go:
```diff
@@ -1530,6 +1544,54 @@ func (b *BlockChain) BlockHashByHeight(blockHeight int32) (*chainhash.Hash, erro
+		return false
+	}
+
+	return !node.status.KnownInvalid()
```
`IsValidHeader` reads `node.status` directly here after all locks are released. Since `SetStatusFlags`/`UnsetStatusFlags` mutate status under the index write lock, this is a data race. Using `b.index.NodeStatus(node).KnownInvalid()` instead of `node.status.KnownInvalid()` would fix it — `NodeStatus` reads under the lock.

yyforyongyu commented on blockchain/process_test.go:
```diff
@@ -0,0 +1,180 @@
+	// Check that the tip is correct.
+	lastHeader := headers[len(headers)-1]
+	lastHeaderHash := lastHeader.BlockHash()
+	tipNode := chain.bestHeader.Tip()
```
Small thing — `tipNode` is captured once here and reused in assertions at lines 134 and 179. If `bestHeader` moved unexpectedly, these checks would still pass against the stale pointer. Re-fetching `chain.bestHeader.Tip()` for each assertion would make the "tip didn't move" checks more robust.

kcalvinalvin commented on netsync/manager_test.go:
```diff
@@ -0,0 +1,239 @@
+	for scanner.Scan() {
+		line := scanner.Text()
+		b, err := hex.DecodeString(line)
+		if err != nil {
```
I'll skip the require package changes so it's easier to diff with other parts of the code change from the outstanding reviews.

kcalvinalvin commented on blockchain/blockindex.go:
```diff
@@ -377,14 +385,15 @@ func newBlockIndex(db database.DB, chainParams *chaincfg.Params) *blockIndex {
+	node, hasBlock := bi.index[*hash]
 	bi.RUnlock()
-	return hasBlock
+	return hasBlock && node.status.HaveData()
```
Addressed. The fix will be included in my next push

kcalvinalvin commented on blockchain/chainio.go:
```diff
@@ -1262,6 +1263,9 @@ func (b *BlockChain) initChainState() error {
 		}
 
+		// Set the lastNode in the blockindex as the bestheader.
+		b.bestHeader.SetTip(lastNode)
```
Good catch. Addressed and will be included in my next push.

kcalvinalvin commented on netsync/manager.go:
```diff
@@ -845,85 +831,64 @@ func (sm *SyncManager) handleBlockMsg(bmsg *blockMsg) {
+	// Fetch more blocks if we're still not caught up to the best header and
+	// don't have any more requestedBlocks.
+	_, lastHeight := sm.chain.BestHeader()
+	if bmsg.block.Height() < lastHeight && len(state.requestedBlocks) == 0 {
```
Addressed. Will be included in my next push

kcalvinalvin commented on netsync/manager.go:
```diff
@@ -969,99 +934,41 @@ func (sm *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
-		sm.headerList.Remove(sm.headerList.Front())
-		log.Infof("Received %v block headers: Fetching blocks",
-			sm.headerList.Len())
 		sm.progressLogger.SetLastLogTime(time.Now())
```
Good point. Address and the fix will be included in my next push

kcalvinalvin commented:
All the outstanding review comments from both reviewers have been addressed.

One thing I'm still working on is https://github.com/btcsuite/btcd/pull/2428#discussion_r2791188069 from YY but that's for adding new tests and isn't a blocker for re-addressing the reviews.

yyforyongyu reviewed (approved):
LGTM🥇

The only comment I have is this [one](https://github.com/btcsuite/btcd/pull/2428/changes#r2920007215), it's non-blocking as I believe it's extremely unlikely to happen on mainnet, tho plausible on testnet or regtest, while I still consider it an edge case, I think it's good to flag it here for future ref.
